### PR TITLE
Refactor backend architecture

### PR DESCRIPTION
**Pull request by Beigelman**
## Reason
This PR has the goal to uncouple various layers of our backend architecture. This is still a work in progress but with one or two other PRs this should be done
The new architecture proposal is the following one:

1. Domain: Responsible for the app domain and all its business logic
2. Infra: In charge of all the communication with external API and Databases
3. App: Control all the use cases for each different domain
4. Presentation: The layer responsible to communicate with all the client in our case using GraphQl to serve our frontend

The test strategy was also changed to suit the new architecture.

## Screenshots

image
image
image

### CODE REVIEW

File: packages/graphql/jest.config.js [hunk 1]
```diff
@@ -1,20 +1,180 @@
-const pkg = require('./package');
+// For a detailed explanation regarding each configuration property, visit:
```
Comment (belgamo) [inline packages/graphql/jest.config.js:1]: Could you clear this file by removing all comments?

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (belgamo) [inline .vscode/launch.json]: I think it's a user settings level. Can you add it to `.gitignore`?

Comment (Beigelman) [reply .vscode/launch.json]: Done!

File: packages/graphql/jest.config.js (hunk 1 above)
Comment (Beigelman) [reply packages/graphql/jest.config.js:1]: Sure

File: packages/graphql/jest.config.js [hunk 3]
```diff
@@ -1,20 +1,13 @@
...
+// https://jestjs.io/docs/en/configuration.html
 
 module.exports = {
-  rootDir: './',
```
Comment (brunohkbx) [inline packages/graphql/jest.config.js:4]: Are these changes really necessary?

File: packages/graphql/package.json [hunk 4]
```diff
@@ -4,19 +4,22 @@
...
+    "start:infra": "docker-compose up -d",
     "test": "jest",
     "test:watch": "jest --watch",
-    "update:schema": "babel-node --extensions \".es6,.js,.es,.jsx,.mjs,.ts,.tsx\" ./scripts/updateSchema.ts"
+    "update:schema": "ts-node-dev ./scripts/updateSchema.ts"
   },
   "dependencies": {
     "@entria/graphql-mongoose-loader": "^4.3.0",
     "@koa/router": "^8.0.6",
+    "awilix": "^4.3.4",
```
Comment (brunohkbx) [inline packages/graphql/package.json:19]: Why these packages were added?

File: packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    authenticateUser,
+    storageProvider,
+    userRepository,
+    facebookRepository,
+    tokenProvider,
+  };
+};
+
+describe('AuthenticateFacebookUser', () => {
+  it('should be able to authenticate', async () => {
```
Comment (brunohkbx) [inline packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts]: Not a big thing but tests shouldn't start with `should` they need to be assertive about what they are doing
https://www.betterspecs.org/#should

File: packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        {
+          id: 'socialId-fake-token',
+          providedBy: 'facebook',
+        },
+      ],
+    });
+    expect(token).toMatch(/(JWT )\w+\.\w+\.\w+/g);
+  });
+
+  it('should be able to authenticate with no user photo', async () => {
```
Comment (brunohkbx) [inline packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts]: Can we break down these different contexts in describe blocks?
Something like:
```js
describe('When the user has no photo', () => {
  it('authenticates it, () => {
   ....
  });
})

Comment (brunohkbx) [reply packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts]: The same for all others below

File: packages/graphql/src/modules/user/app/authenticateFacebookUser.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  expires: number;
+  permissions: string[];
+}
+
+export interface SaveFBResponse {
+  user: User;
+  token: string;
+}
+
+const authenticateFacebookUser = ({
```
Comment (brunohkbx) [inline packages/graphql/src/modules/user/app/authenticateFacebookUser.ts:26]: @Beigelman btw shouldn't this be part of a different PR? I mean, the point here is to refactor the backend architecture right?

**Comment by brunohkbx** [comment]
Nice work here @Beigelman. However, I suggest breaking down the new things in separated PRs in order to make them easier to review. This PR should include only changes regarding the new architecture
Architecture-wise, I wonder if wouldn't be better to stick with the current organization. I mean, this is an open-source project meant to receive help from other people. Applying DDD here definitely increases the learning curve to board new people in the project. Up to you @belgamo

File: packages/graphql/package.json (hunk 4 above)
Comment (Beigelman) [reply packages/graphql/package.json:19]: Awilix is an injection dependency package that helps decouple several layers of the application. It is a great tool for implements some SOLID concepts.

File: packages/graphql/src/modules/user/app/authenticateFacebookUser.ts (hunk 7 above)
Comment (Beigelman) [reply packages/graphql/src/modules/user/app/authenticateFacebookUser.ts:26]: The previous implementation was basically doing everything in one file, since I've isolated different parts of the application I had to refactor this use case as well.

File: packages/graphql/jest.config.js (hunk 3 above)
Comment (Beigelman) [reply packages/graphql/jest.config.js:4]: Yes, due to the nature of the new architecture proposal there were I few things that needed to be changed in order to adjust to this architecture.

File: packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts (hunk 5 above)
Comment (Beigelman) [reply packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts]: Cool, I didn't know that I will address this issue

File: packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts (hunk 6 above)
Comment (Beigelman) [reply packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts:91]: That's a good suggestion, thanks!

File: packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts (hunk 5 above)
Comment (Beigelman) [reply packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts]: Pretty cool this site https://www.betterspecs.org/ by the way!

File: packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts (hunk 6 above)
Comment (Beigelman) [reply packages/graphql/src/modules/user/app/authenticateFacebookUser.spec.ts]: Okay 👍🏼

**Comment by Beigelman** [comment]
> Nice work here @Beigelman. However, I suggest breaking down the new things in separated PRs in order to make them easier to review. This PR should include only changes regarding the new architecture
> Architecture-wise, I wonder if wouldn't be better to stick with the current organization. I mean, this is an open-source project meant to receive help from other people. Applying DDD here definitely increases the learning curve to board new people in the project. Up to you @belgamo

I agree that this could have been broken down into separated PRs and is something that I will have in mind in my next contributions.
Architecture-wise I disagree with you. I find the current architecture very unorganized and difficult to understand. My intention with this PR is to turn the code more readable and easy to contribute.

Thanks for all your comments I will address them soon.

**Comment by belgamo** [comment]
> Nice work here @Beigelman. However, I suggest breaking down the new things in separated PRs in order to make them easier to review. This PR should include only changes regarding the new architecture
> Architecture-wise, I wonder if wouldn't be better to stick with the current organization. I mean, this is an open-source project meant to receive help from other people. Applying DDD here definitely increases the learning curve to board new people in the project. Up to you @belgamo

Nice job @Beigelman and thanks for your review @brunohkbx, I really appreciate it! 

@brunohkbx since that I've started to contribute here I feel the things a little unorganized on the backend side. I mean, a lot of different routines placed together, one class with various responsibilities. Then in this point, I agree with @Beigelman about applying any concept which goal is to improve the organization of the code. 
About the fact it is an opensource project I agree we need to keep things simple as much as possible. But I prefer the architecture a little bit harder to understand over a bad codebase. That's worthwhile.

**Review by belgamo — APPROVED** [review]
Nice job! You just forgot about integration tests and to applying DDD to the adoption concept. Since this PR is big enough, I will map these tasks to be done in another one.
